### ISSUE TITLE
About coustimize jvp

### ISSUE THREAD

**Issue by konglk1203** [google-deepmind/ferminet#82]
Dear authors

I appreciate your beautiful work and I am trying to under the paper 2308.16848 using this repository. Could you explain why we need to custom jvp for the function ``total_energy`` [link](https://github.com/google-deepmind/ferminet/blob/38df5df1959c2d99bfdb05bef5f743f972acf15c/ferminet/loss.py#L237)? I checked the paper carefully and did not find an explaination.

**Comment by jsspencer** [comment]
See Eq 9 and surrounding text in https://journals.aps.org/prresearch/abstract/10.1103/PhysRevResearch.2.033429

Naive application of backprop would require taking a derivative through the kinetic energy operator, resulting in an expensive third-order derivative. [Note that this is not unique to our work, but is a standard result in the VMC community]

**Comment by konglk1203** [comment]
Dear Dr Spencer
Thank you very much for the detailed explaination. I still have the following questions:
1. You said the purpose is to avoid `expensive third-order derivative'. However, I read the reference you send (Eq. 9 in [Pfau 2020](https://journals.aps.org/prresearch/abstract/10.1103/PhysRevResearch.2.033429)) and it was about KFAC optimizer. I am a little confused why the customized jvp function is still called when I use Adam optimizer?
2. I tried to run the code without this customized jvp function, i.e., use the more expensive backward propagation for computing gradient, and use Adam to train it, the model does not even converge (all the other settings are the same except this jvp function). How does this function improve performance?

Thank you so much again

**Comment by jsspencer** [comment]
1. It applies to all optimizers, not just KFAC -- the loss function and its gradient are required in order to optimize any network.
2. You don't say what system and what you mean by not converge but Adam converges much slower than KFAC (if at all, for anything other than small systems) and the custom jvp also does gradient clipping to aid numerical stability. Both these points are discussed in Pfau et al 2020. It would also not surprise me if the naive backward propagation produces numerically sensitive/unstable gradients (due to the determinant) -- even the second derivatives require special handling (Pfau 2020) -- especially if lower precision is used. We (and subsequently others) have shown that second order optimizers are almost always required for efficient optimization of these kinds of networks. Finally I note that when one can, an analytic gradient is almost always faster than one produced via backprop (eg consider taking that gradient through a Fourier transform!)

**Comment by konglk1203** [comment]
Dear Dr Spencer
Thank you so much for the detailed instructions! Please give me some time learning it.
